### DISCUSSION TITLE
Traceback does not totally show the point of failure in my application.

### DISCUSSION THREAD

**Discussion by philipokiokio** [fastapi/fastapi#9475] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Traceback (most recent call last):
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/anyio/streams/memory.py", line 94, in receive
    return self.receive_nowait()
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/anyio/streams/memory.py", line 89, in receive_nowait
    raise WouldBlock
anyio.WouldBlock

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/starlette/middleware/base.py", line 43, in call_next
    message = await recv_stream.receive()
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/anyio/streams/memory.py", line 114, in receive
    raise EndOfStream
anyio.EndOfStream

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/uvicorn/protocols/http/httptools_impl.py", line 372, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/starlette/middleware/base.py", line 68, in __call__
    response = await self.dispatch_func(request, call_next)
  File "/Users/scale/Documents/kakar/./core/middlewares/upload_size_limit.py", line 27, in dispatch
    return await call_next(request)
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/starlette/middleware/base.py", line 46, in call_next
    raise app_exc
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/starlette/middleware/base.py", line 36, in coro
    await self.app(scope, request.receive, send_stream.send)
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/starlette/middleware/cors.py", line 84, in __call__
    await self.app(scope, receive, send)
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/starlette/exceptions.py", line 93, in __call__
    raise exc
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
    await self.app(scope, receive, sender)
  File "/Users/scale/Documents/kakar/venv/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 26, in __call__
    raise dependency_exception
  File "/Users/scale/Documents/kakar/./core/utils/database_utils.py", line 16, in get_db
    yield db
AttributeError: 'Workspace' object has no attribute 'member'
```

### Description

My application follows some tenants of the Repository design, what that means for my application is that, I have service files that call repositories scripts where queries are written. 
`
class CommentsService:
    def __init__(self, db) -> None:
        self.repo = ProjectCommentRepo(ProjectComment, db)
        self.tagged_user = CommentUserRepo(ProjectTaggedUser, db)

`
        
Above is a sample of a service that I used to interact with repos it passes db which is an instance of a session (Sessionmaker instance) that is resolved via Depends in routes.
`
#Example of a repo

class AbstractRepo:
    def __init__(self, model, db: Session) -> None:
        self.model = model
        self.db = db

    @property
    def base_query(self):
        return self.db.query(self.model)

    def create(self, create_dict: dict):
        model_ = self.model(**create_dict)
        self.db.add(model_)
        self.db.commit()
        self.db.refresh(model_)
        return model_

`
However anytime there is an error it is always an extreme sport locating where the issue is since it does not point to the area of failure in any script but returns information about the yield db. Although it sometimes returns the issue below like the sample above. I had to manually locate the point of failure.

### Operating System

macOS

### FastAPI Version

0.79.0

### Python Version

3.9.6

**Comment by Kludex** [comment]
Do you have the same behavior on the latest FastAPI and Python version?

**Reply by philipokiokio** [reply]
No idea. 

I have not upgraded to the latest version of FastAPI (I pretty much stuck with the version I started the project on), so when I use Session Local() directly I get a point of failure included in the traceback log.

**Reply by Kludex** [reply]
Well, it may be a good thing to try to create an MRE, and see if you have the same behavior with the latest FastAPI and Python version.

**Reply by philipokiokio** [reply]
I just did this now. I used the latest version of FastAPI and Python 3.10.10. 

I had the same output.

`
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/anyio/streams/memory.py", line 94, in receive
    return self.receive_nowait()
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/anyio/streams/memory.py", line 89, in receive_nowait
    raise WouldBlock
anyio.WouldBlock

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/starlette/middleware/base.py", line 43, in call_next
    message = await recv_stream.receive()
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/anyio/streams/memory.py", line 114, in receive
    raise EndOfStream
anyio.EndOfStream

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 372, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/starlette/middleware/base.py", line 68, in __call__
    response = await self.dispatch_func(request, call_next)
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/./core/middlewares/upload_size_limit.py", line 27, in dispatch
    return await call_next(request)
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/starlette/middleware/base.py", line 46, in call_next
    raise app_exc
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/starlette/middleware/base.py", line 36, in coro
    await self.app(scope, request.receive, send_stream.send)
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/starlette/middleware/cors.py", line 84, in __call__
    await self.app(scope, receive, send)
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/starlette/exceptions.py", line 93, in __call__
    raise exc
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/starlette/exceptions.py", line 82, in __call__
    await self.app(scope, receive, sender)
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/env/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 26, in __call__
    raise dependency_exception
  File "/Users/scale/Documents/projects/mre/Kakar-Backend/./core/utils/database_utils.py", line 16, in get_db
    yield db
TypeError: delete_refresh_token() missing 1 required positional argument: 'db'


`


this is what my get_db function looks like. (I don't know if that will help)
`
def get_db():
    db = SessionLocal_()
    try:
        yield db

    except Exception as e:
        db.rollback()

    finally:
        db.close()


`

This is an example of a route method, with the DB dependency.


`


@auth.get("/refresh/", status_code=status.HTTP_200_OK)
def get_new_access_token(
    token: str,
    db: Session = Depends(get_db),
):
    resp = AuthService(db).refresh(token)
    return resp

`

**Reply by philipokiokio** [reply]
So I tried something which is probably not the best implementation, and I currently don't know the implication of these because I am still building up my generator knowledge.
`
def get_db():
    db = SessionLocal_()
     try:
      return  db

except Exception as e:
    db.rollback()

finally:
    db.close()
`

So when I did this I saw all the Traceback logs.  Does this implementation work with the same logic of closing every session at the end of the request.

**Reply by philipokiokio** [reply]
Hi, so I experimented with using yield and return. 

I am speaking in context to the get_db function.
I noticed that I don't see point of failure with Yield but I see the exact point of failure with return. 


What can I do to remedy this.

**Comment by philipokiokio** [comment]
This issue with my Traceback was the fact that I had an exception in my get_db function. 


There were some conversation about this on the FastAPI Advanced Question Channel. Here is a link to the discussion that led to the solution. (https://discord.com/channels/772607987558252564/772631851675418624/1103587759618265098)


TL DR solution.
you can comment out the exception when you that you should see the full traceback.



Someone named Doctor suggested that I could stick to the context manager style to yield session. He said that it would call the rollback session and close methods. 


That is stated here (https://discord.com/channels/772607987558252564/772631851675418624/1103610346960330773).
